### PR TITLE
feat(microsoft): add users sync

### PR DESCRIPTION
**Pull request by elbalexandre**
## Description

- add authentication
- add users sync
- add token refresh

## Additional Notes

Still need more tests on local 

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/microsoft/src/inngest/functions/users/sync-users-page.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    const { organisationId, syncStartedAt, tenantId, skipToken, region } = event.data;
+
+    const [organisation] = await db
+      .select({ token: Organisation.token })
+      .from(Organisation)
+      .where(
+        and(
+          eq(Organisation.id, organisationId),
+          eq(Organisation.tenantId, tenantId),
+          eq(Organisation.region, region)
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/functions/users/sync-users-page.ts:51]: Same here, region check is worthless, we are not going to run into uuid collision.

File: apps/microsoft/src/inngest/middlewares/unauthorized-middleware.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                  apiKey: env.ELBA_API_KEY,
+                  baseUrl: env.ELBA_API_BASE_URL,
+                });
+                await db
+                  .delete(Organisation)
+                  .where(
+                    and(
+                      eq(Organisation.id, data.organisationId),
+                      eq(Organisation.tenantId, data.tenantId),
+                      eq(Organisation.region, data.region)
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/middlewares/unauthorized-middleware.ts:52]: same here

File: apps/microsoft/src/inngest/middlewares/unauthorized-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                    )
+                  );
+                await elba.connectionStatus.update({ hasError: true });
+              }
+              return {
+                ...context,
+                result: {
+                  ...result,
+                  error: new NonRetriableError(
+                    `Github return an unauthorized status code for '${fn.name}'`,
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/middlewares/unauthorized-middleware.ts:62]:
```suggestion
                    `Microsoft return an unauthorized status code for '${fn.name}'`,
```

File: apps/microsoft/src/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { NextResponse } from 'next/server';
+import type { NextRequest } from 'next/server';
+import { validateWebhookRequestSignature } from '@elba-security/sdk';
+import { env } from '@/env';
+
+export async function middleware(request: NextRequest) {
+  try {
+    await validateWebhookRequestSignature(request, env.ELBA_WEBHOOK_SECRET);
+  } catch (error) {
+    // TODO: add log
```
Comment (elbalexandre) [inline apps/microsoft/src/middleware.ts:10]: Missing log here

File: apps/microsoft/src/database/client.node.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { drizzle } from 'drizzle-orm/postgres-js';
```
Comment (elbalexandre) [inline apps/microsoft/src/database/client.node.ts:1]: This file can be removed

File: apps/microsoft/package.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    "react": "^18",
+    "react-dom": "^18",
+    "zod": "^3.22.4"
+  },
+  "devDependencies": {
+    "@edge-runtime/vm": "^3.1.7",
+    "@elba-security/eslint-config-custom": "workspace:*",
+    "@elba-security/test-utils": "workspace:*",
+    "@elba-security/tsconfig": "workspace:*",
+    "@microsoft/microsoft-graph-types": "^2.40.0",
```
Comment (elbalexandre) [inline apps/microsoft/package.json:43]: I might not need that rn

**Comment by elbalexandre** [comment]
I have forgot to encode token in database.

File: apps/microsoft/src/inngest/functions/users/sync-users-page.test.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+import * as usersConnector from '@/connectors/users';
+import { db } from '@/database/client';
+import { Organisation } from '@/database/schema';
+import type { MicrosoftUser } from '@/connectors/users';
+import { env } from '@/env';
+import { syncUsersPage } from './sync-users-page';
+
+const organisation = {
+  id: '45a76301-f1dd-4a77-b12f-9d7d3fca3c90',
+  token: 'test-token',
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/functions/users/sync-users-page.test.ts:13]:
```suggestion
  token: encrypt('test-token'),
```

File: apps/microsoft/src/inngest/functions/token/refresh-token.test.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        .where(eq(Organisation.id, organisation.id))
+    ).resolves.toStrictEqual([
+      {
+        token: newToken,
+      },
+    ]);
+
+    expect(authConnector.getToken).toBeCalledTimes(1);
+    expect(authConnector.getToken).toBeCalledWith(organisation.tenantId);
+    expect(crypto.encrypt).toBeCalledTimes(1);
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/functions/token/refresh-token.test.ts:83]: I am not sure it's worse testing encrypt is called.
By testing what is in the db we are already doing the check.

File: apps/microsoft/src/inngest/functions/token/refresh-token.test.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    await expect(result).resolves.toBe(undefined);
+
+    await expect(
+      db
+        .select({ token: Organisation.token })
+        .from(Organisation)
+        .where(eq(Organisation.id, organisation.id))
+    ).resolves.toStrictEqual([
+      {
+        token: newToken,
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/functions/token/refresh-token.test.ts:77]:
```suggestion
        token: encrypt(newToken),
```

File: apps/microsoft/src/inngest/functions/token/refresh-token.test.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    expect(step.sendEvent).toBeCalledTimes(0);
+  });
+
+  test('should update encrypted token and schedule the next refresh', async () => {
+    await db.insert(Organisation).values(organisation);
+    vi.spyOn(authConnector, 'getToken').mockResolvedValue({
+      token: newToken,
+      expiresIn,
+    });
+    vi.spyOn(crypto, 'encrypt').mockResolvedValue(newToken);
```
Comment (elbalexandre) [inline apps/microsoft/src/inngest/functions/token/refresh-token.test.ts:60]:
```suggestion
```
We could not mock this

**Review by lweinhard — APPROVED** [review]

**Pull request merged**